### DISCUSSION TITLE
array_init_elem on dropped element segment

### DISCUSSION THREAD

**Discussion by CharlieTap** [WebAssembly/gc#577] [Q&A] [answered]
At the end of the wast script for array_init_elem there's these two tests, I'm unsure how the second test should pass if the first test drops the element it needs?

```wast
;; init_data/elem with dropped segments traps for non-zero length
(assert_return (invoke "drop_segs"))
(assert_return (invoke "array_init_elem" (i32.const 0) (i32.const 0) (i32.const 0)))
```

**Comment by rossberg** [comment, marked as answer]
Semantically, dropping a segment shrinks its size to zero, so a zero-length bulk op at offset 0 still succeeds (modulo null references). In this case, see the [execution rule](https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-array-mathsf-array-init-elem-x-y), steps 20-21.

This behaviour is inherited from prior bulk ops like table.init and memory.init, though I cannot find corresponding tests for those with a quick search.

The original motivation for this semantics is that the drop check becomes subsumed by the bounds check this way, so that an extra runtime check is avoided. [This PR](https://github.com/WebAssembly/bulk-memory-operations/pull/126) links to some of the history of bulk op boundary conditions.

**Reply by CharlieTap** [reply]
Thanks for the context, that all makes sense
